parser.y

```diff
@@ -7994,9 +7995,26 @@ TableOption:
 	{
 		$$ = &ast.TableOption{Tp: ast.TableOptionStatsPersistent}
 	}
+|	"STATS_AUTO_RECALC" EqOpt LengthNum
+	{
+		n := $3.(uint64)
+		if n != 0 && n != 1 {
+			yylex.AppendError(yylex.Errorf("The value of STATS_AUTO_RECALC must be one of [0|1|DEFAULT]."))
+			return 1
+		}
+		$$ = &ast.TableOption{Tp: ast.TableOptionStatsAutoRecalc, UintValue: n}
+		yylex.AppendError(yylex.Errorf("The STATS_AUTO_RECALC is parsed but ignored by all storage engines."))
+		parser.lastErrorAsWarn()
+	}
+|	"STATS_AUTO_RECALC" EqOpt "DEFAULT"
+	{
+		$$ = &ast.TableOption{Tp: ast.TableOptionStatsAutoRecalc, UintValue: 0}
```

Comment:
@leoppro @lauhg I don't think `DEFAULT` is the same as 0. 

Here is the MySQL 8.0 [documentation](https://dev.mysql.com/doc/refman/8.0/en/create-table.html):
> - STATS_AUTO_RECALC
Specifies whether to automatically recalculate persistent statistics for an InnoDB table. The value DEFAULT causes the persistent statistics setting for the table to be determined by the innodb_stats_auto_recalc configuration option. The value 1 causes statistics to be recalculated when 10% of the data in the table has changed. The value 0 prevents automatic recalculation for this table;

TiDB needs the information to decide whether using `innodb_stats_auto_recalc`(or something similar) or not.

Reply:
@tangenta I agree with you.
@leoppro What do you think?

Reply:
fine, I think we should add a feild named `default` to TableOption

Reply:
could fix `"STATS_SAMPLE_PAGES" EqOpt "DEFAULT"` by the way?

Reply:
Could this be parsed into a `nil` option?

Reply:
@kennytm what do you mean?

Reply:
> fine, I think we should add a feild named `default` to TableOption

I think this is a good way to solve this problem.

Reply:
@lauhg Since `STATS_SAMPLE_PAGES = DEFAULT` is the default the option could just be entirely removed from the AST, so changing this branch to `$$ = nil` should be possible.

Reply:
@kennytm As far as I know, this parser is not only part of TiDB, it will also be used by other projects. In order to be more compatible with mysql, I think it is necessary to know if the default value is explicitly set.

Reply:
@lauhg What project would benefit from knowing that the DEFAULT is explicitly set? In MySQL (8.0.17) even if you explicit write `STATS_AUTO_RECALC = DEFAULT` it will disappear in `SHOW CREATE TABLE`, showing these are semantically equivalent.

```mysql
mysql> create table t (a int) stats_auto_recalc = 0;
Query OK, 0 rows affected (0.06 sec)

mysql> show create table t;
+-------+--------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                         |
+-------+--------------------------------------------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `a` int(11) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_general_ci STATS_AUTO_RECALC=0 |
+-------+--------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> drop table t;
Query OK, 0 rows affected (0.02 sec)
```

```mysql
mysql> create table t (a int) stats_auto_recalc = default;
Query OK, 0 rows affected (0.01 sec)

mysql> show create table t;
+-------+------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                     |
+-------+------------------------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `a` int(11) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_general_ci |
+-------+------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> drop table t;
Query OK, 0 rows affected (0.00 sec)
```

Reply:
@kennytm 
In `create table` case, these are semantically equivalent.
But in `alter table` case, these are not equivalent.
For example:
```sql
create table t (a int) stats_auto_recalc = 1;
alter table t stats_auto_recalc = default;
```
If parser parses `stats_auto_recalc = default` to `nil`, it would be equivalent to
`alter table t`.
Apparently, it changes semantics.